mediatek: add AdTran SmartRG Bonanza Peak series

dangowrt opened the pull request:
The Adtran SDG-8600 ("Bonanza Peak") series are a couple of MT7986-powered Wi-Fi 6 residential gateway, access point and mesh router products. 

https://www.adtran.com/en/products-and-services/residential-solutions/mesh-wi-fi-gateways-and-satellites/service-delivery-gateways/sdg-8600-series

All of them come with an eMMC to boot from, are powered via USB-C and got a USB 3.0 type-A port. All of them got a Dialog (Renesas) DA14531 Bluetooth module connected via UART. If the device was previously running stock firmware, the BT chip's internal flash has been loaded with firmware and it can be attached using hciattach when using OpenWrt.

 * **SDG-8612 - Dual-band RJ-45 gateway**
   1x 2.5G MaxLinear PHY for WAN port
   4x 1GE LAN ports via MT7531 switch

 * **SDG-8614 - Dual-band SFP gateway**
   SFP cage with up to 2.5G speed
   4x 1GE LAN ports via MT7531 switch

 * **SDG-8622 - Tri-band mesh router**
   2x 2.5G MaxLinear PHY
   The MT7986 2G and 5G are used as 2G and 5G high band.
   There’s a MT7915 PCIe card for 5G low band.

 * **SDG-8632 - Tri-band mesh router with 6 GHz**
   2x 2.5G MaxLinear PHY
   The MT7986 serves the 2G and 6G bands.
   There’s a MT7915 PCIe card for 5G.

dangowrt committed:
base-files: add mmc_get_mac_ascii function

Similar to the *_get_mac_binary function, also split the common parts
off mtd_get_mac_ascii into new get_mac_ascii function and introduce
mmc_get_mac_ascii which uses it.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: import driver for Adtran SmartRG RGBW LED

Import driver for I2C-connected HolTek MCU controlling the RGBW LED
found in Adtran SmartRG devices.

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

dangowrt committed:
mediatek: add support for Adtran SmartRG Bonanza Peak series

The Bonanza Peak series are a couple of MT7986-powered 2.5 GBit/s
Wi-Fi 6 residential gateway, access point and mesh router products.

All of them come with an eMMC to boot from, are powered via USB-C and
got a USB 3.0 type-A port. All of them got a Dialog (Renesas) DA14531
Bluetooth module connected via UART. If the device was previously
running stock firmware, the BT chip's internal flash has been loaded
with firmware and it can be attached using hciattach when using
OpenWrt.

SOC: MediaTek MT7986A
RAM: 2 GiB DDR4
eMMC: 8 GiB
Bluetooth: BLE5 (DA14531)
Serial: 3.3V level, 115200 8n1 on 4-pin connector

 * SDG-8612 - Dual-band RJ-45 gateway
   2x 2.5G MaxLinear PHY for WAN port
   3x 1GE LAN ports via MT7531 switch

 * SDG-8614 - Dual-band SFP gateway
   1x SFP cage with up to 2.5G speed
   1x 2.5G MaxLinear PHY for LAN port
   3x 1GE LAN ports via MT7531 switch

 * SDG-8622 - Tri-band mesh router
   2x 2.5G MaxLinear PHY
   The MT7986 2G and 5G are used as 2G and 5G high band.
   There’s a MT7915 PCIe card for 5G low band.

 * SDG-8632 - Tri-band mesh router with 6 GHz
   2x 2.5G MaxLinear PHY
   The MT7986 serves the 2G and 6G bands.
   There’s a MT7915 PCIe card for 5G.

Installation via U-Boot serial console:
0. setup TFTP server with IP 192.168.1.10/24, place initramfs image
renamed to openwrt.XXX where XXX is the internal product number:
 SDG-8612: XXX = 412
 SDG-8614: XXX = 414
 SDG-8622: XXX = 422
 SDG-8632: XXX = 432

1. connect to the serial console and power on the device.
   Interrupt the bootloader by pressing 'st'
2. setenv boot_mode openwrt ; saveenv
3. run boot1
   Load firmware via TFTP and write to flash
4. run boot2
   Now OpenWrt initramfs should boot
5. upload sysupgrade.bin via scp to /tmp
6. sysupgrade

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

blocktrron commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+		compatible = "maxlinear,gpy211", "ethernet-phy-ieee802.3-c45";
+		reg = <5>;
+
+		mxl,led-drive-vdd;
```
Is this a leftover or is the patch for this property missing?

dangowrt replied:
Oh yes, Adtran uses a custom patch to setup LEDs, the property inverts the output of the LED. I'll remove it until we have proper support for LEDs on the GPY211 PHY.

blocktrron replied:
I'll have a look at implementing LED control in an upstream compatible way.

dangowrt replied:
@blocktrron I've started working on that but didn't have time to finish it:
https://github.com/dangowrt/linux/commit/96d12cfb7b3b237e4058136a07e79c880da6cf4c

cmonroe replied:
For whatever reason the LEDs were driven by VDD in this design instead of the PHY LED pins. The "mxl,led-drive-vdd" property was a hack to set this via the PHY_LED register (0.27 15:12) as seen in the [datasheet.](https://assets.maxlinear.com/web/documents/617810_gpy211b1vc_gpy211c0vc_ds_rev1.4.pdf)

blocktrron replied:
@dangowrt thanks, working on it

abajk replied:
@dangowrt I want to convert the last devices using this hack. Do you know what colors LEDs have?

dangowrt replied:
Ironically I those devices were the whole reason why I wrote the LED support for MaxLinear PHYs, so I had that part ready. However, it then turned out that the MT7531 switch cannot control LEDs individually on the PHY-level, but settings are shared among all 5 PHY LEDs -- one would have to use the switch-level GPIO controller to control them individually (as GPIOs). Hence I was planning to support LEDs of the MT7531 on the switch-level (ie. only offload netdev led trigger in case all used LEDs have identical settings), and then change all those Adtran devices to have proper LED support for all Ethernet LEDs of both, MaxLinear GPY211 PHYs as well as MT7531 switch.

Meanwhile we can apply this:
```diff
diff --git a/target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi b/target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi
index deab53dc45..9da5e06874 100644
--- a/target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi
+++ b/target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi
@@ -220,8 +220,33 @@
 		compatible = "maxlinear,gpy211", "ethernet-phy-ieee802.3-c45";
 		reg = <5>;
 
-		mxl,led-drive-vdd;
-		mxl,led-config = <0x30 0x40 0x80 0x0>;
+		leds {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			led@0 {
+				reg = <0>;
+				color = <LED_COLOR_ID_AMBER>;
+				function = LED_FUNCTION_LAN;
+				active-high;
+			};
+
+			led@1 {
+				reg = <1>;
+				color = <LED_COLOR_ID_GREEN>;
+				function = LED_FUNCTION_LAN;
+				function-enumerator = <0>;
+				active-high;
+			};
+
+			led@2 {
+				reg = <2>;
+				color = <LED_COLOR_ID_GREEN>;
+				function = LED_FUNCTION_LAN;
+				function-enumerator = <0>;
+				active-high;
+			};
+		};
 	};
 
 	phy6: ethernet-phy@6 {
@@ -229,8 +254,33 @@
 		compatible = "maxlinear,gpy211", "ethernet-phy-ieee802.3-c45";
 		reg = <6>;
 
-		mxl,led-drive-vdd;
-		mxl,led-config = <0x30 0x40 0x80 0x0>;
+		leds {
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			led@0 {
+				reg = <0>;
+				color = <LED_COLOR_ID_AMBER>;
+				function = LED_FUNCTION_WAN;
+				active-high;
+			};
+
+			led@1 {
+				reg = <1>;
+				color = <LED_COLOR_ID_GREEN>;
+				function = LED_FUNCTION_WAN;
+				function-enumerator = <0>;
+				active-high;
+			};
+
+			led@2 {
+				reg = <2>;
+				color = <LED_COLOR_ID_GREEN>;
+				function = LED_FUNCTION_WAN;
+				function-enumerator = <0>;
+				active-high;
+			};
+		};
 	};
 };
 ```

abajk replied:
It will be hard to add triggers for two leds of the same color. Are all LEDs available in each model of router?

blocktrron commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+		compatible = "maxlinear,gpy211", "ethernet-phy-ieee802.3-c45";
+		reg = <6>;
+
+		mxl,led-drive-vdd;
```
same

blocktrron commented on target/linux/mediatek/dts/mt7986a-smartrg-SDG-8632.dts:
```diff
@@ -0,0 +1,38 @@
+};
+
+&wifi {
+	ieee80211-freq-limit = <2400000 2500000>, <5945000 7125000>;
```
Interesting, is this a Dual-Band Radio which can operate on 2.4GHz or 6GHz? I've never seen something like this.

dangowrt replied:
Yes, they are using MT7986A to drive a MT7976A (not MT7976C like on many other boards), and that is a Wi-Fi 6E DBDC frontend which can do 6 GHz.

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-SDG-8612.dts:
```diff
@@ -0,0 +1,76 @@
+};
+
+&mdio {
+	switch: switch@0 {
```
```suggestion
	switch: switch@1f {
```

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-SDG-8614.dts:
```diff
@@ -0,0 +1,105 @@
+};
+
+&mdio {
+	switch: switch@0 {
```
```suggestion
	switch: switch@1f {
```

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+		pinctrl-names = "default";
+		pinctrl-0 = <&button_pins>;
+
+		reset {
```
```suggestion
		button-reset {
```

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+		compatible = "gpio-leds";
+
+		mux_sel {
+			label = "mux_sel";
```
Labels are deprecated.

dangowrt replied:
Thanks for the review.
I've applied your comments, replace LED labels and represented GPIO-controlled mux as gpio-hog instead of LED with default state.

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+		reg = <0x30>;
+
+		system_red {
+			label = "red";
```
Labels are deprecated.

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+		compatible = "srg,sysled";
+		reg = <0x30>;
+
+		system_red {
```
```suggestion
		led@1 {
```

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+			reg = <1>;
+		};
+
+		sys_status: system_green {
```
```suggestion
		sys_status: led@2 {
```

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+			linux,default-trigger = "timer";
+		};
+
+		system_blue {
```
```suggestion
		led@3 {
```

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+		sys_status: system_green {
+			label = "green";
+			reg = <2>;
+			linux,default-trigger = "timer";
```
Add comment here also about timer without delay parameters?

dangowrt replied:
removed timer, it was a left-over from stock firmware.

raenye commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,491 @@
+			reg = <3>;
+		};
+
+		system_white {
```
```suggestion
		led@4 {
```

dangowrt committed:
kernel: hack: support inverted LEDs on MaxLinear GPY211 PHY

Add downstream DT property to setup the PHY LEDs of the MaxLinear
GPY211 PHY in such way that the VDD of the LED is driven by the SoC
pin rather than the GND (which is the default).

Signed-off-by: Daniel Golle <daniel@makrotopia.org>

DragonBluep commented on target/linux/mediatek/dts/mt7986a-smartrg-SDG-8612.dts:
```diff
@@ -0,0 +1,76 @@
+	switch: switch@1f {
+		compatible = "mediatek,mt7531";
+		reg = <31>;
+		reset-gpios = <&pio 5 GPIO_ACTIVE_HIGH>;
```
Hi! I found mt7531 driver reset has some issues. Hope you can have a look: https://github.com/openwrt/openwrt/pull/13125#discussion_r1414764232

DragonBluep commented on target/linux/mediatek/dts/mt7986a-smartrg-SDG-8614.dts:
```diff
@@ -0,0 +1,124 @@
+			function = "sfp";
+			function-enumerator = <1>;
+			gpios = <&pio 16 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
```
`default-state = "off";` should be removed as OFF is the default behavior in Linux kernel. However, u-boot still requires it.

DragonBluep commented on target/linux/mediatek/dts/mt7986a-smartrg-bonanza-peak.dtsi:
```diff
@@ -0,0 +1,481 @@
+
+&ssusb {
+	vusb33-supply = <&reg_3p3v>;
+	vbus-supply = <&reg_5v>;
```
The fixed voltage regulator seems useless. https://github.com/torvalds/linux/blob/c527f5606aa545233a4d2c6d5c636ed82b8633ef/Documentation/devicetree/bindings/usb/mediatek%2Cmtk-xhci.yaml#L181
